File: dp_revision/lecture_5/p1.cpp

// memorization

class Solution {
public:
    int solve(vector<int>& nums,int l,int r,vector<vector<int>>& dp)
    {
        if(l+1 == r)
            return 0;
        if(dp[l+1][r+1]!=-1)
            return dp[l+1][r+1];
        int ans=0;
        for(int i=l+1;i<r;i++)
        {
            int temp1=solve(nums,l,i,dp);
            int temp2=solve(nums,i,r,dp);
            int lv=1;
            int rv=1;
            if(l!=-1)
                lv=nums[l];
            if(r!=nums.size())
                rv=nums[r];
            
            int temp_ans=temp1+temp2+(nums[i] * lv * rv);
            ans=max(ans,temp_ans);
        }
        dp[l+1][r+1]=ans;
        return ans;
    }
    int maxCoins(vector<int>& nums) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        vector<vector<int>> dp(nums.size()+2,vector<int>(nums.size()+2,-1));
        return solve(nums,-1,nums.size(),dp);
    }
};

// tabulation

class Solution {
public:
    int maxCoins(vector<int>& nums) {
        if(nums.size()==0)
            return 0;
        
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        
        vector<int> arr(nums.size()+2);
        arr[0]=1;
        arr[arr.size()-1]=1;
        for(int i=0;i<nums.size();i++)
            arr[i+1]=nums[i];
                
        vector<vector<int>> dp(arr.size(),vector<int>(arr.size(),0));
        for(int g=2;g<arr.size();g++)
        {
            for(int i=0;i<arr.size()-g;i++)
            {
                int j=i+g;
                for(int k=1;k<g;k++)
                {
                    int left=dp[i][i+k];
                    int right=dp[i+k][j];
                    dp[i][j]=max(dp[i][j] ,left + right + arr[i] * arr[i+k] * arr[j]);
                }
            }
        }
        return dp[0][dp[0].size()-1];
    }
};


class Solution {
public:
    int maxCoins(vector<int>& nums) {
        if(nums.size()==0)
            return 0;
        
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        
        vector<int> arr(nums.size()+2);
        arr[0]=1;
        arr[arr.size()-1]=1;
        for(int i=0;i<nums.size();i++)
            arr[i+1]=nums[i];
        
        
        vector<vector<int>> dp(arr.size(),vector<int>(arr.size(),0));
        for(int g=2;g<arr.size();g++)
        {
            for(int i=0;i<arr.size()-g;i++)
            {
                int j=i+g;
                for(int k=1;k<g;k++)
                {
                    int left=dp[i][i+k];
                    int right=dp[i+k][j];
                    dp[i][j]=max(dp[i][j] ,left + right + (i-1>=0?nums[i-1]:1) * nums[i+k-1] * (j-1<nums.size()?nums[j-1]:1));
                }
            }
        }
        return dp[0][dp[0].size()-1];
    }
};

File: lecture_8/p1.java
public class p1 {
    public static int solve(int arr[],int i,int j,int strg[][])
    {
        if(i+1==j)
            return 0;
        if(strg[i][j]!=0)
            return strg[i][j];
        int ans=Integer.MAX_VALUE;
        for(int k=i+1;k<j;k++)
        {
            ans=Math.min(ans,solve(arr,i,k,strg)+solve(arr,k,j,strg)+arr[i]*arr[j]*arr[k]);
        }
        strg[i][j]=ans;
        return ans;
    }
    public static int minScoreTriangulation(int[] arr) {
        if(arr.length==0)
            return 0;
        int strg[][]=new int[arr.length][arr.length];
        return solve(arr,0,arr.length-1,strg);
    }

    // tabulation
    public static int minScoreTriangulation_02(int[] arr) {
        if(arr.length==0)
            return 0;
        int strg[][]=new int[arr.length][arr.length];
        for(int g=2;g<strg.length;g++)
        {
            for(int i=0;i<strg.length-g;i++)
            {
                int j=i+g;
                if(g==2)
                {
                    strg[i][j]=arr[i]*arr[i+1]*arr[i+2];
                }
                else
                {
                    int ans=Integer.MAX_VALUE;
                    for(int k=i+1;k<j;k++)
                    {
                        int left=strg[i][k];
                        int right=strg[k][j];
                        int tcost=left+right+arr[i]*arr[j]*arr[k];
                        if(tcost<ans)
                            ans=tcost;
                    }
                    strg[i][j]=ans; 
                }
                
            }
        }
        return strg[0][arr.length-1];
    } 
    public static void main(String args[])
    {
        int arr[]={1,2,3};
        System.out.println(minScoreTriangulation(arr));
        System.out.println(minScoreTriangulation_02(arr));
    }
}